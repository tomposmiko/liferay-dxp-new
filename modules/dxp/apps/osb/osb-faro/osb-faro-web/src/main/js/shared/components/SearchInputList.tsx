import BaseSelect from './BaseSelect';
import getCN from 'classnames';
import Label from 'shared/components/Label';
import React, {useEffect, useRef, useState} from 'react';

interface ISearchInputListProps extends React.HTMLAttributes<HTMLElement> {
	clearOnAdd: boolean;
	containerClass?: string;
	dataSourceFn: (inputValue: string) => Promise<any>;
	disabled: boolean;
	itemRenderer?: () => React.ReactNode;
	items: any[];
	onItemsChange: (items: any[]) => void;
	placeholder?: string;
}

const SearchInputList: React.FC<ISearchInputListProps> = ({
	className,
	clearOnAdd = false,
	containerClass,
	dataSourceFn,
	disabled,
	itemRenderer,
	items,
	onItemsChange,
	placeholder = ''
}) => {
	const [focus, setFocus] = useState<boolean>(false);
	const [query, setQuery] = useState<string>('');

	const _baseSelectRef = useRef<any>();

	useEffect(() => {
		if (_baseSelectRef && !!items.length) {
			_baseSelectRef.current.focus();
		}
	}, [items]);

	const handleQueryChange = (value: string): void => {
		setQuery(value);
	};

	const handleAddItem = (item: any): void => {
		if (item) {
			onItemsChange([...items, item]);
		}

		if (clearOnAdd) {
			setQuery('');
		}
	};

	const handleRemoveItem = (index: number): void => {
		onItemsChange([...items.slice(0, index), ...items.slice(index + 1)]);
	};

	return (
		<div
			className={getCN(
				'search-input-list-root input-group-item',
				className,
				{disabled, focus}
			)}
		>
			<div className='form-control form-control-tag-group d-flex'>
				{items &&
					!!items.length &&
					items.map((item, i) => (
						<Label
							display='secondary'
							index={i}
							key={i}
							onRemove={handleRemoveItem}
						>
							{item}
						</Label>
					))}

				<BaseSelect
					className='flex-grow-1'
					containerClass={getCN(
						'search-input-list-container',
						containerClass
					)}
					dataSourceFn={dataSourceFn}
					disabled={disabled}
					emptyInputOnInactive
					inputSize='sm'
					inputValue={query}
					inset
					itemRenderer={itemRenderer}
					onBlur={() => setFocus(false)}
					onFocus={() => setFocus(true)}
					onInputValueChange={handleQueryChange}
					onSelect={handleAddItem}
					placeholder={placeholder}
					ref={_baseSelectRef}
				/>
			</div>
		</div>
	);
};

export default SearchInputList;
