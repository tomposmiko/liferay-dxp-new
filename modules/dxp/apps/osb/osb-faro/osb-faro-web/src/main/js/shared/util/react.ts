import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
