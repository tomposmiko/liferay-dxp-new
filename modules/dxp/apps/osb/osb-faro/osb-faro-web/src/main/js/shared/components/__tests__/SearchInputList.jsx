import React from 'react';
import SearchInputList from '../SearchInputList';
import {cleanup, render} from '@testing-library/react';

jest.unmock('react-dom');

describe('SearchInputList', () => {
	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(<SearchInputList items={[]} />);

		expect(container).toMatchSnapshot();
	});
});
