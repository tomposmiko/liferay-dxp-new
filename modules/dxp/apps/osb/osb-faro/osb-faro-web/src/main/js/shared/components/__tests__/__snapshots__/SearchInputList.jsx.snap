// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInputList should render 1`] = `
<div>
  <div
    class="search-input-list-root input-group-item"
  >
    <div
      class="form-control form-control-tag-group d-flex"
    >
      <div
        aria-haspopup="true"
        class="input-group base-select-input-root select-input-root flex-grow-1 inset"
      >
        <div
          class="input-group-item"
        >
          <input
            autocomplete="off"
            class="input-root form-control input-group-inset input-group-inset-after form-control-sm"
            placeholder=""
            type="text"
            value=""
          />
          <div
            class="input-group-inset-item input-group-inset-item-after"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-bottom icon-root"
              role="presentation"
            >
              <use
                xlink:href="#caret-bottom"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
